metrics/examples/prometheus/install/strimzi-service-monitor.yaml

```diff
@@ -0,0 +1,123 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: kafka-service-monitor
+  labels:
+    app: strimzi
+spec:
+  selector:
+    matchLabels:
+      strimzi.io/name: my-cluster-kafka-bootstrap
```

Comment:
This label should probably not be here.

Reply:
Correct me if I am worng, but Kafka bootstrap service, which has the metrics we want to scrape has these labels 
```
  labels:
    strimzi.io/cluster: my-cluster
    strimzi.io/kind: Kafka
    strimzi.io/name: my-cluster-kafka-bootstrap
```
so we beed this label or add `app: strimzi` label to this service.

Reply:
So ... this is there to tell Prometheus which services should the metrics be collected from? That was not clear to me?  Ok.

However: 
* `strimzi.io/name: my-cluster-kafka-bootstrap` would only capture Kafka metrics for one particular cluster. You would need to make it super clear that this needs to be changed. 
* What about Zookeeper, Connect or Mirror Maker? 
* Using the `strimzi.io/kind` labels seems to be less dependent on the cluster name ... and AFAIK we should have similar labels for Connect and MirrorMaker. 
* On the other hand `app: strimzi` is IMHO set only on the CO and would not be useful right now, or?

Reply:
I think that we would need different `ServiceMonitor` resources for Kafka, Zookeeper, Connect and Mirror Maker.

Using `kind` or `name` depends if we want one Prometheus server scraping metrics for more clusters or not (in different namespaces).

The `app: strimzi` label to this service is needed because it's the way how the previous `Prometheus` resource knows of which services it has to take care of.
It has the following section:

```yaml
serviceMonitorSelector:
    matchLabels:
      app: strimzi
```

Reply:
Right. So I reviewed the docs twice already but these things do not seem obvious from it. So that goes probably back to my general comment that this is too detailed to be understandable for someone who doesn't know it.

> The app: strimzi label to this service is needed because it's the way how the previous Prometheus resource knows of which services it has to take care of.

Which service are you talking about here? Our services do not have `app: strimzi` labels. Also, what is _previous Prometheus resource_? Is that the old Prometheus deployment?

> I think that we would need different ServiceMonitor resources for Kafka, Zookeeper, Connect and Mirror Maker.

Then they should be probably in here as well?

Reply:
> Which service are you talking about here? Our services do not have app: strimzi labels. Also, what is previous Prometheus resource? Is that the old Prometheus deployment?

Sorry, "service" is too much overloaded :-) I meant the `ServiceMonitor` kind where we are having this discussion.
Even "previous" use was bad ... I meant the `kind: Prometheus` resources defined in the `prometheus.yaml` in this PR.

Reply:
Sorry, I still don't think I understand. So the YAML has this selector:

```yaml
spec:
  selector:
    matchLabels:
      strimzi.io/name: my-cluster-kafka-bootstrap
      app: strimzi
```

Stanislav said that the `strimzi.io/name: my-cluster-kafka-bootstrap` is used to tell prometheus to scrape the metrics through a service with this label. (at least that is how I understood it) But the `app: strimzi` label is not part of the labels on that service. That is confusing me because I would expect this to not match reliably every Kafka cluster.

Reply:
So...
The app:strimzi label on ServiceMonitor is needed to say to kind:Prometheus which ServiceMonitor resources to take care of. 
The name labels instead in the ServiceMonitor needs to define the scraping configuration so which Kube service to scrape for metrics. 
Prometheus takes care of ServiceMonitor resources matching app:strimzi label and the ServiceMonitor helps to define the scraping for kube service with the name label. 
Now getting out of the train! ;)

Reply:
> The app:strimzi label on ServiceMonitor is needed to say to kind:Prometheus which ServiceMonitor resources to take care of.

But how does `app: strimzi` in `ServiceMonitor.spec.selector.matchLabels` tell `Prometheus` to use this `ServiceMonitor`? Just to make it clear ... are we talking about the same label? I'm talking about the one on line 11. Not the one in `metadata.labels`.

Reply:
Let me try to clarify ...
The `prometheus.yaml`, describing `Prometheus` resource, starting at line 56 has:

```yaml
serviceMonitorSelector:
    matchLabels:
      app: strimzi
```

so `Prometheus.spec.serviceMonitorSelector.matchLabels` which matches the label in the `ServiceMonitor` (file `strimzi-service-monitor.yaml`) at line 6:

```yaml
labels:
    app: strimzi
```

so `ServiceMonitor.metadata.labels`.

This is the way how Prometheus which are the ServiceMonitor(s) to take care of.
Then, we have `ServiceMonitor.spec.selector.matchLabels` which matches the labels of the Kubernetes native `Service` for which scraping the metrics.

Reply:
I think the easiest way how to make it clear is to put comments to these lines & mention that in the docs.

Reply:
Right ... that explains all the other labels I didn't complained about. So back to my original point ... this file, line 11 ... why is `app:strimzi` here? It is not on our services by default. The user would need to specifically request such label through the Kafka custom resource, or?

Reply:
Agree, It should not be there.